// @hubspot/ui-extensions/dist/coreComponents.js
import { createRemoteReactComponent } from '@remote-ui/react';
const Alert = createRemoteReactComponent('Alert');
const Button = createRemoteReactComponent('Button');
const ButtonRow = createRemoteReactComponent('ButtonRow');
const Card = createRemoteReactComponent('Card');
const DescriptionList = createRemoteReactComponent('DescriptionList');
const DescriptionListItem = createRemoteReactComponent('DescriptionListItem');
const Divider = createRemoteReactComponent('Divider');
const EmptyState = createRemoteReactComponent('EmptyState');
const ErrorState = createRemoteReactComponent('ErrorState');
const Form = createRemoteReactComponent('Form');
const Heading = createRemoteReactComponent('Heading');
const Image = createRemoteReactComponent('Image');
const Input = createRemoteReactComponent('Input');
const Textarea = createRemoteReactComponent('Textarea');
const LoadingSpinner = createRemoteReactComponent('LoadingSpinner');
const ProgressBar = createRemoteReactComponent('ProgressBar');
const Select = createRemoteReactComponent('Select');
const Tag = createRemoteReactComponent('Tag');
const Text = createRemoteReactComponent('Text');
const Tile = createRemoteReactComponent('Tile');
const ToggleGroup = createRemoteReactComponent('ToggleGroup');
const Stack = createRemoteReactComponent('Stack');
const StatisticsItem = createRemoteReactComponent('StatisticsItem');
const Statistics = createRemoteReactComponent('Statistics');
const StatisticsTrend = createRemoteReactComponent('StatisticsTrend');
const Table = createRemoteReactComponent('Table');
const TableFooter = createRemoteReactComponent('TableFooter');
const TableCell = createRemoteReactComponent('TableCell');
const TableRow = createRemoteReactComponent('TableRow');
const TableBody = createRemoteReactComponent('TableBody');
const TableHeader = createRemoteReactComponent('TableHeader');
const TableHead = createRemoteReactComponent('TableHead');
const Link = createRemoteReactComponent('Link');
const NumberInput = createRemoteReactComponent('NumberInput');
const Box = createRemoteReactComponent('Box');
export { Alert, Box, Button, ButtonRow, Card, DescriptionList, DescriptionListItem, Divider, EmptyState, ErrorState, Form, Heading, Image, Input, Textarea, LoadingSpinner, Link, ProgressBar, Select, Stack, Statistics, StatisticsItem, StatisticsTrend, Table, TableFooter, TableCell, TableRow, TableBody, TableHeader, TableHead, Tag, Text, Tile, ToggleGroup, NumberInput, };
