// @hubspot/ui-extensions-dev-server/build.js
const { build } = require('vite');
const { ROLLUP_OPTIONS } = require('./constants');

async function buildAllExtensions(config, outputDir) {
  const extensionKeys = Object.keys(config);
  for (let i = 0; i < extensionKeys.length; ++i) {
    const { data } = config[extensionKeys[i]];

    await buildSingleExtension({
      file: data?.module.file,
      outputFileName: data?.output,
      outputDir,
      emptyOutDir: i === 0,
    });
  }
}

async function buildSingleExtension({
  file,
  outputFileName,
  outputDir,
  emptyOutDir = true,
  plugins = { rollup: [], vite: [] },
  minify = false,
  root = process.cwd(), // This is the vite default, so using that as our default
}) {
  await build({
    root,
    define: {
      'process.env.NODE_ENV': JSON.stringify(
        process.env.NODE_ENV || 'production'
      ),
    },
    build: {
      lib: {
        entry: file,
        name: outputFileName,
        formats: ['iife'],
        fileName: () => outputFileName,
      },
      rollupOptions: {
        ...ROLLUP_OPTIONS,
        plugins: [...(ROLLUP_OPTIONS.plugins || []), ...plugins?.rollup],
      },
      outDir: outputDir,
      emptyOutDir,
      minify,
      plugins: plugins?.vite,
    },
  });
}

module.exports = {
  buildAllExtensions,
  buildSingleExtension,
};
