// @hubspot/ui-extensions-dev-server/cli.js
const logger = require('./logger');
const commandLineArgs = require('command-line-args');
const commandLineUsage = require('command-line-usage');

function parseArgs() {
  const mainDefinitions = [{ name: 'command', defaultOption: true }];
  const mainOptions = commandLineArgs(mainDefinitions, {
    stopAtFirstUnknown: true,
  });
  const argv = mainOptions._unknown || [];
  const DEV_MODE = mainOptions.command === 'dev';
  const BUILD_MODE = mainOptions.command === 'build';

  const optionDefinitions = [
    { name: 'port', alias: 'p', type: Number },
    { name: 'extension', alias: 'e', type: String },
    { name: 'help', alias: 'h', type: Boolean },
  ];

  const options = commandLineArgs(
    optionDefinitions,
    DEV_MODE || BUILD_MODE ? { argv } : {}
  );

  return { DEV_MODE, BUILD_MODE, ...options };
}

function showHelp(OUTPUT_DIR) {
  const sections = [
    {
      header: 'HubSpot UI Extensions Local Dev Server',
      content: `Used for local development of HubSpot extensions.  Built assets can be found in the ${OUTPUT_DIR} directory`,
    },
    {
      header: 'Available Commands',
      content: [
        { name: 'dev', summary: 'starts the local development server' },
        { name: 'build', summary: 'runs a build of your extensions' },
      ],
    },
    {
      header: 'Options',
      optionList: [
        {
          name: 'extension',
          alias: 'e',
          typeLabel: '{underline file}',
          description:
            'The extension entrypoint file to build or start local development for',
        },
        {
          name: 'help',
          alias: 'h',
          description: 'Print this usage guide.',
        },
      ],
    },
  ];
  const usage = commandLineUsage(sections);
  logger.info(usage);
  process.exit(0);
}

module.exports = {
  parseArgs,
  showHelp,
};
